src/server/chromium/crNetworkManager.ts

```diff
@@ -542,3 +563,133 @@ const errorReasons: { [reason: string]: Protocol.Network.ErrorReason } = {
+
+  constructor() {
+  }
+
+  requestWillBeSent(event: Protocol.Network.requestWillBeSentPayload) {
+    const info = this._requests.get(event.requestId);
+    if (info) {
+      // This is redirect.
+      if (!event.redirectResponse?.connectionId) {
+        // Bail out early for responses without connectionId.
+        // They won't be getting extra info, so there is nothing to track.
+        this._stopTracking(event.requestId);
+        return;
+      }
+      info.responseReceived = true;
+    } else {
+      this._requests.set(event.requestId, {
+        requestId: event.requestId,
+        responseReceived: false,
+        responseReceivedExtraInfo: [],
+        responses: [],
+      });
+    }
+  }
+
+  responseReceived(event: Protocol.Network.responseReceivedPayload) {
+    const info = this._requests.get(event.requestId);
+    if (!info)
+      return;
+
+    // Bail out early for responses without connectionId.
+    // They won't be getting extra info, so there is nothing to track.
+    if (!event.response.connectionId)
+      this._stopTracking(event.requestId);
+  }
+
+  responseReceivedExtraInfo(event: Protocol.Network.responseReceivedExtraInfoPayload) {
+    const info = this._requests.get(event.requestId);
+    if (!info)
+      return;
+    info.responseReceivedExtraInfo.push(event);
+    this._patchResponseHeaders(info, info.responseReceivedExtraInfo.length - 1);
+    this._checkFinished(info);
+  }
+
+  processResponse(requestId: string, response: network.Response, wasFulfilled: boolean) {
+    // We are not interested in ExtraInfo tracking for fulfilled requests, our Blink
+    // headers are the ones that contain fulfilled headers.
+    if (wasFulfilled) {
+      this._stopTracking(requestId);
+      return;
+    }
+
+    const info = this._requests.get(requestId);
+    if (!info)
+      return;
+    response.setWillReceiveExtraHeaders();
+    info.responses.push(response);
+    this._patchResponseHeaders(info, info.responses.length - 1);
+  }
+
+  loadingFinished(event: Protocol.Network.loadingFinishedPayload) {
+    const info = this._requests.get(event.requestId);
+    if (!info)
+      return;
+    info.loadingFinished = event;
+    this._checkFinished(info);
+  }
+
+  loadingFailed(event: Protocol.Network.loadingFailedPayload) {
+    const info = this._requests.get(event.requestId);
+    if (!info)
+      return;
+    info.loadingFailed = event;
+    this._checkFinished(info);
+  }
+
+  private _patchResponseHeaders(info: RequestInfo, index: number) {
+    const response = info.responses[index];
+    const extraInfo = info.responseReceivedExtraInfo[index];
+    if (response && extraInfo)
+      response.extraHeadersReceived(headersObjectToArray(extraInfo.headers));
+  }
+
+  private _checkFinished(info: RequestInfo) {
+    if (!info.loadingFinished && !info.loadingFailed)
+      return;
+
+    // Loading finished, check that we have all ExtraInfo events in place.
+    if (!info.responseReceived) {
```

Comment:
According to my other comment where "responseReceived" actually means "has redirect", it seems we are not waiting for the extra info on responses without redirects.

Reply:
Yeah, because "responseReceived" means "responseReceived", not "has redirect" as you suggested.